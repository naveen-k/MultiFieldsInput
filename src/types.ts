type STYLES = {
  container?: Record<string, unknown>
  label?: Record<string, unknown>
  input?: Record<string, unknown>
  error?: Record<string, unknown>
  inputContainer?: Record<string, unknown>
}

export type INPUT = {
  maxLength: number
  placeholder: string
}

export interface Props {
  autoFocus?: boolean
  inputs: INPUT[]
  error?: string
  label?: string
  name: string
  onBlur: (values: FIELDS) => void
  onChange?: (values: FIELDS) => void
  value?: string
  styles?: STYLES
}

export type FIELDS = {
  name: string
  value: string
}
