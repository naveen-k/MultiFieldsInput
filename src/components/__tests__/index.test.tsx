import React from 'react'
import { render } from '@testing-library/react'
import user from '@testing-library/user-event'

import { INPUT } from '../../types'
import MultiFieldsInput from '../'

describe('MultiFieldsInput', () => {
  const inputs: INPUT[] = [
    {
      maxLength: 2,
      placeholder: '00',
    },
    {
      maxLength: 2,
      placeholder: '00',
    },
    {
      maxLength: 2,
      placeholder: '00',
    },
  ]

  const defaultProps = {
    label: 'Sort Code',
    name: 'sortCode',
    inputs: inputs,
    onBlur: jest.fn(),
    onChange: jest.fn(),
  }

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('Renders correctly', () => {
    const { container } = render(
      <MultiFieldsInput {...defaultProps} value="202020" />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('Calls onBlur and onChange and sends the correct value', () => {
    const { getByTestId } = render(<MultiFieldsInput {...defaultProps} />)

    const field1 = getByTestId(/sortCode0/i)
    user.type(field1, '10')

    expect(defaultProps.onChange).toHaveBeenCalledTimes(3)
    expect(defaultProps.onChange).toHaveBeenCalledWith({
      name: 'sortCode',
      value: '10',
    })
    expect(defaultProps.onBlur).toHaveBeenCalledTimes(1)

    const field2 = getByTestId(/sortCode1/i)
    user.type(field2, '20')

    expect(defaultProps.onChange).toHaveBeenCalledTimes(5)
    expect(defaultProps.onChange).toHaveBeenCalledWith({
      name: 'sortCode',
      value: '1020',
    })
    expect(defaultProps.onBlur).toHaveBeenCalledTimes(2)
  })

  it('Error is displayed when passed as props', () => {
    const { getByText } = render(
      <MultiFieldsInput {...defaultProps} error={'Invalid input'} />
    )
    expect(getByText(/Invalid input/i)).toBeInTheDocument()
  })

  it('Displays passed styling', () => {
    const styles = {
      label: {
        display: 'block',
      },
      input: {
        color: '#2D3748',
      },
      error: {
        fontSize: '13px',
      },
    }

    const { getByTestId, getByText } = render(
      <MultiFieldsInput
        {...defaultProps}
        error={'Invalid input'}
        styles={styles}
      />
    )

    expect(getByText(/sort code/i)).toHaveStyle(styles.label)
    expect(getByTestId(/sortCode0/i)).toHaveStyle(styles.input)
    expect(getByText(/Invalid input/i)).toHaveStyle(styles.error)
  })
})
