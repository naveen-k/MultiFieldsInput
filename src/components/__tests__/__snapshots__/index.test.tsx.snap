// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiFieldsInput Renders correctly 1`] = `
<div>
  <label
    for="sortCode"
  >
    Sort Code
  </label>
  <div>
    <input
      data-testid="sortCode0"
      maxlength="2"
      name="sortCode0"
      placeholder="00"
      type="text"
      value="20"
    />
    <input
      data-testid="sortCode1"
      maxlength="2"
      name="sortCode1"
      placeholder="00"
      type="text"
      value="20"
    />
    <input
      data-testid="sortCode2"
      maxlength="2"
      name="sortCode2"
      placeholder="00"
      type="text"
      value="20"
    />
  </div>
</div>
`;
